src/fp.rs

```diff
@@ -0,0 +1,299 @@
+        let mut lo: u64;
+        let mut cc: u64;
+
+        // Integer multiplication
+        result = mul64(x[0], y[0]);
+        carry = result.0;
+        zz[0] = result.1;
+        result = mul64(x[0], y[1]);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        zz[1] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        zz[2] = result.0;
+
+        result = mul64(x[1], y[0]);
+        hi = result.0;
+        lo = result.1;
+        result = add64(zz[1], lo, 0);
+        zz[1] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        carry = result.0;
+
+        result = mul64(x[1], y[1]);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        lo = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        hi = result.0;
+        result = add64(zz[2], lo, 0);
+        zz[2] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        zz[3] = result.0;
+
+        // Reduction
+        let w = self.mu.wrapping_mul(zz[0]);
+        result = mul64(p[0], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(zz[0], lo, 0);
+        zz[0] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        carry = result.0;
+
+        result = mul64(p[1], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        lo = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        hi = result.0;
+        result = add64(zz[1], lo, 0);
+        zz[1] = result.0;
+        cc = result.1;
+        result = add64(zz[2], hi, cc);
+        zz[2] = result.0;
+        cc = result.1;
+        result = add64(zz[3], 0, cc);
+        zz[3] = result.0;
+
+        let w = self.mu.wrapping_mul(zz[1]);
+        result = mul64(p[0], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(zz[1], lo, 0);
+        zz[1] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        carry = result.0;
+
+        result = mul64(p[1], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        lo = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        hi = result.0;
+        result = add64(zz[2], lo, 0);
+        zz[2] = result.0;
+        cc = result.1;
+        result = add64(zz[3], hi, cc);
+        zz[3] = result.0;
+
+        (zz[2] as u128) | ((zz[3] as u128) << 64)
+    }
+
+    pub fn inv(&self, x: u128) -> u128 {
+        let exp = self.p - 2;
+        let mut t = self.elem(1);
+        for i in (0..128).rev() {
+            t = self.mul(t, t);
+            if (exp >> i) & 1 != 0 {
+                t = self.mul(t, x);
+            }
+        }
+        t
+    }
+
+    pub fn neg(&self, x: u128) -> u128 {
+        self.sub(self.elem(0), x)
+    }
+
+    pub fn elem(&self, x: u128) -> u128 {
```

Comment:
you may define a new type for the internal representation.
e.g.  using the following functions when you move from/to Montgomery representation
newElt(x: u128) -> fpElt 
getU128(x: fpElt ) ->  u128

Reply:
That will be handled at a higher level: see https://github.com/cjpatton/libprio-rs/pull/1/files#diff-1895006a15ebe7506aa57e9a6c46a3b1ad2aae83002a21969a6a7827631b912fR114